File: helix-runtime/apps/ios/Sources/Email/EmailModels.swift
import Foundation

/// Represents an email message
struct Email: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let userId: String
    let from: EmailAddress
    let to: [EmailAddress]
    let cc: [EmailAddress]?
    let bcc: [EmailAddress]?
    let subject: String
    let body: String
    let htmlBody: String?
    let isRead: Bool
    let isStarred: Bool
    let threadId: String
    let messageId: String?
    let inReplyTo: String?
    let labels: [String]
    let attachmentCount: Int
    let timestamp: Date
    let receivedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, userId, from, to, cc, bcc, subject, body, htmlBody
        case isRead = "is_read"
        case isStarred = "is_starred"
        case threadId = "thread_id"
        case messageId = "message_id"
        case inReplyTo = "in_reply_to"
        case labels
        case attachmentCount = "attachment_count"
        case timestamp
        case receivedAt = "received_at"
    }
}

/// Email address with name and email
struct EmailAddress: Codable, Equatable, Hashable {
    let name: String?
    let email: String

    var displayName: String {
        name ?? email
    }
}

/// Draft email being composed
struct EmailDraft: Identifiable, Codable, Equatable {
    var id: String = UUID().uuidString
    var to: [EmailAddress] = []
    var cc: [EmailAddress] = []
    var bcc: [EmailAddress] = []
    var subject: String = ""
    var body: String = ""
    var inReplyTo: String?
    var attachments: [EmailAttachment] = []
    var timestamp: Date = Date()
}

/// Email attachment
struct EmailAttachment: Identifiable, Codable, Equatable {
    let id: String
    let filename: String
    let mimeType: String
    let size: Int
    let url: URL?
}

/// Email search filter options
struct EmailSearchFilter: Codable, Equatable {
    var query: String = ""
    var from: String?
    var to: String?
    var subject: String?
    var label: String?
    var isRead: Bool?
    var isStarred: Bool?
    var startDate: Date?
    var endDate: Date?
    var hasAttachments: Bool?

    var isEmpty: Bool {
        query.isEmpty && from == nil && to == nil && subject == nil && label == nil &&
            isRead == nil && isStarred == nil && startDate == nil && endDate == nil && hasAttachments == nil
    }
}

/// Email analytics statistics
struct EmailAnalytics: Codable, Equatable {
    let totalEmails: Int
    let unreadCount: Int
    let starredCount: Int
    let spamCount: Int
    let draftCount: Int
    let storageUsedBytes: Int
    let storageQuotaBytes: Int
    let mostActiveSenders: [EmailSenderStats]
    let emailFrequencyByDay: [EmailFrequencyData]

    enum CodingKeys: String, CodingKey {
        case totalEmails = "total_emails"
        case unreadCount = "unread_count"
        case starredCount = "starred_count"
        case spamCount = "spam_count"
        case draftCount = "draft_count"
        case storageUsedBytes = "storage_used_bytes"
        case storageQuotaBytes = "storage_quota_bytes"
        case mostActiveSenders = "most_active_senders"
        case emailFrequencyByDay = "email_frequency_by_day"
    }
}

/// Statistics about email senders
struct EmailSenderStats: Codable, Equatable {
    let email: String
    let displayName: String?
    let messageCount: Int
    let lastMessageDate: Date

    enum CodingKeys: String, CodingKey {
        case email
        case displayName = "display_name"
        case messageCount = "message_count"
        case lastMessageDate = "last_message_date"
    }
}

/// Email frequency data point
struct EmailFrequencyData: Codable, Equatable {
    let date: Date
    let count: Int
}

/// Pagination info for email lists
struct EmailPaginationInfo: Codable, Equatable {
    let totalCount: Int
    let pageSize: Int
    let pageNumber: Int
    let hasMore: Bool

    enum CodingKeys: String, CodingKey {
        case totalCount = "total_count"
        case pageSize = "page_size"
        case pageNumber = "page_number"
        case hasMore = "has_more"
    }
}

File: helix-runtime/apps/android/app/src/main/java/com/helix/email/EmailModels.kt
package com.helix.email

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Date

// MARK: - Email Models

@Serializable
data class Email(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("account_id")
    val accountId: String,
    val from: EmailAddress,
    val to: List<EmailAddress>,
    val cc: List<EmailAddress> = emptyList(),
    val bcc: List<EmailAddress> = emptyList(),
    val subject: String,
    val body: String,
    @SerialName("html_body")
    val htmlBody: String? = null,
    @SerialName("is_read")
    val isRead: Boolean = false,
    @SerialName("is_starred")
    val isStarred: Boolean = false,
    val labels: List<String> = emptyList(),
    @SerialName("thread_id")
    val threadId: String? = null,
    @SerialName("message_id")
    val messageId: String,
    @SerialName("in_reply_to")
    val inReplyTo: String? = null,
    val attachments: List<EmailAttachment> = emptyList(),
    val headers: Map<String, String> = emptyMap(),
    @SerialName("received_at")
    val receivedAt: String,
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
    @SerialName("deleted_at")
    val deletedAt: String? = null,
    @SerialName("is_deleted")
    val isDeleted: Boolean = false,
) {
    val displayDate: String
        get() = receivedAt.takeLast(10) // Simple formatting: YYYY-MM-DD
}

@Serializable
data class EmailAddress(
    val name: String? = null,
    val email: String,
) {
    val displayName: String
        get() = name ?: email
}

@Serializable
data class EmailDraft(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("account_id")
    val accountId: String,
    val to: List<EmailAddress>,
    val cc: List<EmailAddress> = emptyList(),
    val bcc: List<EmailAddress> = emptyList(),
    val subject: String,
    val body: String,
    val attachments: List<EmailAttachment> = emptyList(),
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
)

@Serializable
data class EmailAttachment(
    val id: String,
    val filename: String,
    @SerialName("content_type")
    val contentType: String,
    @SerialName("size_bytes")
    val sizeBytes: Long,
    val url: String? = null,
)

@Serializable
data class EmailSearchFilter(
    val query: String = "",
    val from: String? = null,
    val to: String? = null,
    @SerialName("before_date")
    val beforeDate: String? = null,
    @SerialName("after_date")
    val afterDate: String? = null,
    @SerialName("has_attachment")
    val hasAttachment: Boolean? = null,
    @SerialName("is_read")
    val isRead: Boolean? = null,
    @SerialName("is_starred")
    val isStarred: Boolean? = null,
    val labels: List<String> = emptyList(),
)

@Serializable
data class EmailAnalytics(
    @SerialName("total_emails")
    val totalEmails: Int,
    @SerialName("unread_count")
    val unreadCount: Int,
    @SerialName("starred_count")
    val starredCount: Int,
    @SerialName("total_size_bytes")
    val totalSizeBytes: Long,
    @SerialName("sender_stats")
    val senderStats: List<EmailSenderStats> = emptyList(),
    @SerialName("frequency_data")
    val frequencyData: EmailFrequencyData? = null,
)

@Serializable
data class EmailSenderStats(
    val email: String,
    val name: String? = null,
    @SerialName("email_count")
    val emailCount: Int,
    @SerialName("last_email_at")
    val lastEmailAt: String,
)

@Serializable
data class EmailFrequencyData(
    @SerialName("by_day_of_week")
    val byDayOfWeek: Map<String, Int> = emptyMap(),
    @SerialName("by_hour_of_day")
    val byHourOfDay: Map<String, Int> = emptyMap(),
    @SerialName("by_sender")
    val bySender: Map<String, Int> = emptyMap(),
)

@Serializable
data class EmailPaginationInfo(
    @SerialName("current_page")
    val currentPage: Int,
    @SerialName("page_size")
    val pageSize: Int,
    @SerialName("total_count")
    val totalCount: Int,
    @SerialName("has_more")
    val hasMore: Boolean,
)

// MARK: - Error Handling

sealed class EmailError : Exception() {
    object ServiceUnavailable : EmailError()
    object InvalidEmailData : EmailError()
    object EmailNotFound : EmailError()
    object UnauthorizedAccess : EmailError()
    data class NetworkError(val message: String) : EmailError()
    data class DecodingError(val message: String) : EmailError()
    object UnknownError : EmailError()

    val localizedMessage: String
        get() = when (this) {
            ServiceUnavailable -> "Email service is unavailable. Please try again later."
            InvalidEmailData -> "Invalid email data provided."
            EmailNotFound -> "Email not found."
            UnauthorizedAccess -> "You don't have permission to access this email."
            is NetworkError -> "Network error: ${this.message}"
            is DecodingError -> "Failed to decode email data: ${this.message}"
            UnknownError -> "An unknown error occurred."
        }
}
